benchmark/thermalfluid.jl

```diff
@@ -0,0 +1,268 @@
+using Polynomials: fit
+
+struct Sink; end
+function (::Sink)(inlet)
+    always!(inlet.T - inlet.T_instream)
+end
+
+struct FluidPort
+    p::Float64
+    m::Float64
+    T::Float64
+    T_instream::Float64
+end
+FluidPort() = FluidPort(ntuple(_->continuous(), 4)...)
+
+function connect!(
+    inner::NTuple{N,FluidPort},
+    outer::NTuple{M,FluidPort},
+) where {N,M}
+    if N == 2 && M == 0
+        (a, b) = inner
+        always!(a.T_instream - b.T) # stream
+        always!(a.T - b.T_instream) # stream
+        always!(a.m + b.m) # flow
+    elseif N == 1 && M == 1
+        a = only(inner)
+        b = only(outer)
+        always!(a.T_instream - b.T_instream) # stream
+        always!(a.T - b.T) # stream
+        always!(a.m - b.m) # flow
+    else @assert false "unimplemented" end
+    always!(a.p - b.p) # potential
+
+    return nothing
+end
+
+struct HeatPort
+    Q::Float64
+    T::Float64
+end
+HeatPort() = HeatPort(ntuple(_->continuous(), 2)...)
+function connect!(
+    inner::NTuple{N,HeatPort},
+    outer::NTuple{M,HeatPort},
+) where {N,M}
+    if N == 2 && M == 0
+        (a, b) = inner
+        always!(a.T - b.T)
+        always!(a.Q + b.Q)
+    else @assert false "unimplemented" end
+
+    return nothing
+end
+
+@kwdef struct Source
+    p_feed::Float64 = 100000.0
+end
+
+function (this::Source)(outlet)
+    flow(t) = 2.75
+    T(t) = (t > 12 * 3600) * 56.0 + 12.0
+
+    t = sim_time()
+
+    m_flow = continuous()
+
+    always!(m_flow - flow(t))
+    always!(outlet.m + m_flow)
+    always!(outlet.p - this.p_feed)
+    always!(outlet.T - T(t))
+end
+
+@kwdef struct CircularWall{Ne, Nn}
+    T0::Float64 = 0.
+    d::Float64  = 0.05
+    dx::Float64 = 0.01
+
+    t_layer::NTuple{Ne,Float64} = (0.002,)
+    cp::NTuple{Ne, Float64} = (500.,)
+    ρ::NTuple{Ne, Float64}  = (7850.,)
+    λ::NTuple{Ne, Float64}  = (50.,)
+end
+
+function Cn_circular_wall_inner(d, D, cp, ρ)
+    C = pi / 4 * (D^2 - d^2) * cp * ρ
+    return C / 2
+  end
+
+function Cn_circular_wall_outer(d, D, cp, ρ)
+    C = pi / 4 * (D^2 - d^2) * cp * ρ
+    return C / 2
+end
+
+function Ke_circular_wall(d, D, λ)
+    2 * pi * λ / log(D / d)
+end
+
+function (this::CircularWall{Ne, Nn})(inner_heatport, outer_heatport) where {Ne, Nn}
+    Tn = ntuple(_->continuous(), Nn)
+    Qe = ntuple(_->continuous(), Ne)
+
+    foreach(T->initial!(T - this.T0), Tn)
+
+    dn = (this.d, (this.d .+ 2.0 .* cumsum(this.t_layer))...)
+    Cn = ((Cn_circular_wall_inner.(dn[1:Ne], dn[2:Nn], this.cp, this.ρ) .* this.dx)..., 0.)
+    Cn = Cn .+ (0., (Cn_circular_wall_outer.(dn[1:Ne], dn[2:Nn], this.cp, this.ρ) .* this.dx)...)
+    Ke = Ke_circular_wall.(dn[1:Ne], dn[2:Nn], this.λ) .* this.dx
+
+    always!(inner_heatport.T - Tn[1])
+    always!(outer_heatport.T - Tn[Nn])
+
+    ntuple(Ne) do i
+        always!(Qe[i] - Ke[i] * (Tn[i] - Tn[i+1]))
+        nothing
+    end
+
+    always!(ddt(Tn[1]) * Cn[1] - (inner_heatport.Q - Qe[1]))
+    ntuple(Nn-2) do j
```

Comment:
Does this work now? When I tried this, I quickly ran into the manual unroll limit of ~15 in Base

Reply:
`Nn` is `3` ;). There is another `N-3` ntuple below, but that'll need to be replaced with some log-tree generated function anyway - I'll cross that bridge when we get there.